File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/CS/Queries/QueriesControl.cs
using System;
using System.Linq;
using System.Collections.Generic;
using CrmDataContext;
using System.Configuration;
using System.Windows.Forms;

namespace LinqDemo.Queries {

  public partial class QueriesControl : BaseControl {

    public QueriesControl()
      : base("Queries") {

      InitializeComponent();
      InitializeCombobox();
    }

    private List<CrmLinqQuery> queries;

    private List<string> captions = new List<string>(new string[] { 
      "All companies",
      "Orders for German companies",
      "All orders and companies",
      "Orders count for company",
      "Companies with existing Orders",
      "Average discount for company",
      "Maximum discount for company",
      "Orders grouped by year"
    });

    internal override void OpenClick() {

      dataGridView.DataSource = null;
      tbQuery.Text = "";
      InitializeQueries();
      RefreshQuery();

      base.OpenClick();
    }

    internal override void CloseClick() {

      crmDataContext = null;
      dataGridView.DataSource = null;
      tbQuery.Text = "";

      base.CloseClick();
    }

    private void InitializeQueries() {

      queries = new List<CrmLinqQuery>()
      {
        new CrmLinqQuery(
          captions[0],
          "var q = from c in CrmDataContext.Companies \r\n" +
          "    select c;", 
          AllCompanies()),
        new CrmLinqQuery(
          captions[1], 
          "var q = from company in CrmDataContext.Companies\r\n" +
          "    join order in CrmDataContext.Orders\r\n" +
          "    on company.CompanyID equals order.CompanyID\r\n" +
          "    where company.Country == \"Germany\"\r\n" +
          "    select new { company.Country, company.CompanyName, order.OrderID, order.OrderDate };",
          OrdersForGermanCompanies()),
        new CrmLinqQuery(
          captions[2], 
          "var q = from order in CrmDataContext.Orders\r\n" +
          "    orderby order.Company.CompanyName\r\n" +
          "    select new { order.Company.CompanyName, order.OrderID, order.OrderDate, order.Discount};",
          AllOrdersAndCompanies()),
        new CrmLinqQuery(
          captions[3],
          "var q = from c in CrmDataContext.Companies\r\n" +
          "    orderby c.CompanyName\r\n" +
          "    select new { c.CompanyName, OrderCount = c.Company_Orders.Count };",
          OrdersCountForCompany()),
        new CrmLinqQuery(
          captions[4],
          "var q = from c in CrmDataContext.Companies\r\n" +
          "    join o in CrmDataContext.Orders on c.CompanyID equals o.Company.CompanyID into g\r\n" +
          "    where g.Count() > 0\r\n" +
          "    orderby c.CompanyName\r\n" +
          "    select new { c.CompanyName, OrderCount = g.Count() }",
          CompaniesWithOrders()),
        new CrmLinqQuery(
          captions[5],
          "var q = from c in CrmDataContext.Companies where c.Company_Orders.Count > 0\r\n" +
          "    orderby c.CompanyName\r\n" +
          "    select new { c.CompanyName,\r\n" +
          "      AverageDiscount = c.Orders.Average(o => o.Discount) };",
          AverageDiscountForCompany()),
        new CrmLinqQuery(
          captions[6],
          "var q = from c in CrmDataContext.Companies\r\n" +
          "    orderby c.CompanyName\r\n" +
          "    from o in CrmDataContext.Orders\r\n" +
          "    where c.CompanyID == o.Company.CompanyID\r\n" +
          "    group new { c, o } by c.CompanyName into g\r\n" +
          "    select new { g.Key, MaxDiscount = g.Max(gr => gr.o.Discount) };",
          MaxDiscountForCompany()),
        new CrmLinqQuery(
          captions[7],
          "var q = from o in CrmDataContext.Orders\r\n" +
          "    orderby o.OrderDate descending\r\n" +
          "    select new { o.OrderDate.Year, o.OrderID, o.OrderDate,\r\n" +
          "      o.Company.CompanyName };",
          OrdersGroupedByYear())
      };

    }

    private IQueryable AllCompanies() {

      var q = from c in CrmDataContext.Companies
              select c;
      return q;
    }

    private IQueryable OrdersForGermanCompanies() {

      var q = from company in CrmDataContext.Companies
              join order in CrmDataContext.Orders on company.CompanyID equals order.CompanyID
              where company.Country == "Germany"
              select new { company.Country, company.CompanyName, order.OrderID, order.OrderDate };
      return q;
    }

    private IQueryable AllOrdersAndCompanies() {

      var q = from order in CrmDataContext.Orders
              orderby order.Company.CompanyName
              select new { order.Company.CompanyName, order.OrderID, order.OrderDate, order.Discount};
      return q;
    }

    private IQueryable OrdersCountForCompany() {

      var q = from c in CrmDataContext.Companies
              orderby c.CompanyName
              select new { c.CompanyName, OrderCount = c.Company_Orders.Count };
      return q;
    }

    private IQueryable CompaniesWithOrders() {

      var q = from c in CrmDataContext.Companies
              join o in CrmDataContext.Orders on c.CompanyID equals o.Company.CompanyID into g
              where g.Count() > 0
              orderby c.CompanyName
              select new { c.CompanyName, OrderCount = g.Count() };
      return q;
    }

    private IQueryable AverageDiscountForCompany() {

      var q = from c in CrmDataContext.Companies
              where c.Company_Orders.Count > 0
              orderby c.CompanyName
              select new
              {
                c.CompanyName,
                AverageDiscount = c.Company_Orders.Average(o => o.Discount)
              };
      return q;
    }

    private IQueryable MaxDiscountForCompany() {

      var q = from c in CrmDataContext.Companies
              from o in CrmDataContext.Orders
              where c.CompanyID == o.Company.CompanyID
              group new { c, o } by c.CompanyName into g
              orderby g.Key
              select new { g.Key, MaxDiscount = g.Max(gr => gr.o.Discount) };
      return q;
    }

    private IQueryable OrdersGroupedByYear() {

      var q = from o in CrmDataContext.Orders
              orderby o.OrderDate descending
              select new { o.OrderDate.Year, o.OrderID, o.OrderDate, o.Company.CompanyName };
      return q;
    }

    private void InitializeCombobox() {

      cbQuery.Items.Clear();
      foreach (string s in captions) {
        cbQuery.Items.Add(s);
      }
      cbQuery.SelectedIndex = 0;
    }

    private void RefreshQuery() {

      dataGridView.DataSource = null;
      tbQuery.Text = String.Empty;
      if (crmDataContext == null)
        return;

      string caption = cbQuery.SelectedItem as string;
      CrmLinqQuery query = null;
      foreach (CrmLinqQuery item in queries)
        if (item.Name == caption) {
          query = item;
          break;
        }

      if (query != null) {
        dataGridView.DataSource = query.Query;
        tbQuery.Text = query.Description;
      }
    }

    internal override void ControlsEnabled(bool state) {
     
      cbQuery.Enabled = state;
    }

    private void cbQuery_SelectedIndexChanged(object sender, EventArgs e) {

      try {
        RefreshQuery();
      }
      catch(Exception ex)
      {
        MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      }
    }

    private void dataGridView_DataError(object sender, System.Windows.Forms.DataGridViewDataErrorEventArgs e) {

      if (e.Exception is ArgumentException)
        MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      else
        MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
    }
  }

  internal class CrmLinqQuery {

    public CrmLinqQuery(string name, string description, IQueryable query) {

      Name = name;
      Description = description;
      Query = query;
    }

    public string Name { get; set; }
    public string Description { get; set; }
    public IQueryable Query { get; set; }

    public override string ToString() {

      return Name;
    }
  }
}

File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/VB/Queries/QueriesControl.vb
Public Class QueriesControl
    Inherits BaseControl

    Public Sub New()
        MyBase.New("Queries")
        InitializeComponent()
        InitializeCombobox()
    End Sub

    Private queries As List(Of CrmLinqQuery)
    Private captions As List(Of String) = New List(Of String)(New String() _
                                                              {"All companies", _
                                                              "Orders for German companies", _
                                                              "All orders and companies", _
                                                              "Orders count for company", _
                                                              "Companies with existing Orders", _
                                                              "Average discount for company", _
                                                              "Maximum discount for company", _
                                                              "Orders grouped by year"})

    Friend Overrides Sub OpenClick()
        DataGridView.DataSource = Nothing
        tbQuery.Text = ""
        InitializeQueries()
        RefreshQuery()
        MyBase.OpenClick()
    End Sub

    Friend Overrides Sub CloseClick()
        DataGridView.DataSource = Nothing
        tbQuery.Text = ""
        MyBase.CloseClick()
    End Sub

    Private Sub InitializeQueries()
        queries = New List(Of CrmLinqQuery)

        queries.Add(New CrmLinqQuery(Me.captions(0), _
            "Dim q = From c In myCrmDataContext.Companies _" & ChrW(13) & ChrW(10) & _
            "    Select c", _
            Me.AllCompanies))

        queries.Add(New CrmLinqQuery(Me.captions(1), _
        "Dim q = From company In myCrmDataContext.Companies _" & ChrW(13) & ChrW(10) & _
        "    Join order In myCrmDataContext.Orders _" & ChrW(13) & ChrW(10) & _
        "    On company.CompanyID Equals order.CompanyID _" & ChrW(13) & ChrW(10) & _
        "    Where company.Country = ""Germany"" _" & ChrW(13) & ChrW(10) & _
        "    Select company.Country, company.CompanyName, order.OrderID, order.OrderDate", _
        Me.OrdersForGermanCompanies))

        queries.Add(New CrmLinqQuery(Me.captions(2), _
            "Dim q = From order In myCrmDataContext.Orders _" & ChrW(13) & ChrW(10) & _
            "    Order By order.Company.CompanyName _" & ChrW(13) & ChrW(10) & _
            "    Select order.OrderDate, order.Discount, order.Company.CompanyName", _
        Me.AllOrdersAndCompanies))

        queries.Add(New CrmLinqQuery(Me.captions(3), _
            "Dim q = From c In myCrmDataContext.Companies _" & ChrW(13) & ChrW(10) & _
            "    Order By c.CompanyName _" & ChrW(13) & ChrW(10) & _
            "    Select c.CompanyName, OrderCount = c.Company_Orders.Count", _
        Me.OrdersCountForCompany))

        queries.Add(New CrmLinqQuery(Me.captions(4), _
            "Dim q = From c In myCrmDataContext.Companies _" & ChrW(13) & ChrW(10) & _
            "    Group Join o In myCrmDataContext.Orders _" & ChrW(13) & ChrW(10) & _
            "    Where (g.Count() > 0) _" & ChrW(13) & ChrW(10) & _
            "    Order By c.CompanyName _" & ChrW(13) & ChrW(10) & _
            "    Select c.CompanyName, OrderCount = g.Count()", _
        Me.CompaniesWithOrders))

        queries.Add(New CrmLinqQuery(Me.captions(5), _
            "Dim q = From c In myCrmDataContext.Companies Where c.Company_Orders.Count() > 0 _" & ChrW(13) & ChrW(10) & _
            "    Order By c.CompanyName _" & ChrW(13) & ChrW(10) & _
            "    Select c.CompanyName, _" & ChrW(13) & ChrW(10) & _
            "    AverageDiscount = c.Company_Orders.Average(Function(o) o.Discount)", _
        Me.AverageDiscountForCompany))

        queries.Add(New CrmLinqQuery(Me.captions(6), _
            "Dim q = From c In myCrmDataContext.Companies _" & ChrW(13) & ChrW(10) & _
            "    From o In myCrmDataContext.Orders() _" & ChrW(13) & ChrW(10) & _
            "    Where c.CompanyID = o.Company.CompanyID _" & ChrW(13) & ChrW(10) & _
            "    Group c, o By key = c.CompanyName Into MaxDiscount = Max(o.Discount) _" & ChrW(13) & ChrW(10) & _
            "    Order By key _" & ChrW(13) & ChrW(10) & _
            "    Select key, MaxDiscount", _
        Me.MaxDiscountForCompany))

        queries.Add(New CrmLinqQuery(Me.captions(7), _
            "Dim q = From o In myCrmDataContext.Orders _" & ChrW(13) & ChrW(10) & _
            "    Order By o.OrderDate Descending _" & ChrW(13) & ChrW(10) & _
            "    Select o.OrderDate.Year, o.OrderID, o.OrderDate, o.Company.CompanyName", _
        Me.OrdersGroupedByYear))
    End Sub

    Private Function AllCompanies() As IQueryable
        Dim q = From c In myCrmDataContext.Companies _
              Select c
        Return q
    End Function

    Private Function OrdersForGermanCompanies() As IQueryable
        Dim q = From company In myCrmDataContext.Companies _
          Join order In myCrmDataContext.Orders _
          On company.CompanyID Equals order.CompanyID _
          Where company.Country = "Germany" _
          Select company.Country, company.CompanyName, order.OrderID, order.OrderDate
        Return q
    End Function

    Private Function AllOrdersAndCompanies() As IQueryable
        Dim q = From order In myCrmDataContext.Orders _
                Order By order.Company.CompanyName _
              Select order.Company.CompanyName, order.OrderID, order.OrderDate, order.Discount
        Return q
    End Function

    Private Function OrdersCountForCompany() As IQueryable
        Dim q = From c In myCrmDataContext.Companies _
                Order By c.CompanyName _
              Select c.CompanyName, OrderCount = c.Company_Orders.Count()
        Return q
    End Function

    Private Function CompaniesWithOrders() As IQueryable
        Dim q = From c In myCrmDataContext.Companies _
              Group Join o In myCrmDataContext.Orders _
              On c.CompanyID Equals o.Company.CompanyID Into g = Group _
              Where (g.Count() > 0) _
              Order By c.CompanyName _
          Select c.CompanyName, OrderCount = g.Count()
        Return q
    End Function

    Private Function AverageDiscountForCompany() As IQueryable
        Dim q = From c In myCrmDataContext.Companies Where c.Company_Orders.Count() > 0 _
                Order By c.CompanyName _
              Select c.CompanyName, _
        AverageDiscount = c.Company_Orders.Average(Function(o) o.Discount)
        Return q
    End Function

    Private Function MaxDiscountForCompany() As IQueryable
        Dim q = From c In myCrmDataContext.Companies _
              From o In myCrmDataContext.Orders() _
              Where c.CompanyID = o.Company.CompanyID _
              Group c, o By key = c.CompanyName Into MaxDiscount = Max(o.Discount) _
              Order By key _
          Select key, MaxDiscount
        Return q
    End Function

    Private Function OrdersGroupedByYear() As IQueryable
        Dim q = From o In myCrmDataContext.Orders _
              Order By o.OrderDate Descending _
              Select o.OrderDate.Year, o.OrderID, o.OrderDate, o.Company.CompanyName
        Return q
    End Function

    Private Sub InitializeCombobox()
        Me.cbQuery.Items.Clear()
        Dim text As String
        For Each text In Me.captions
            Me.cbQuery.Items.Add(text)
        Next
        Me.cbQuery.SelectedIndex = 0
    End Sub

    Private Sub RefreshQuery()
        dataGridView.DataSource = Nothing
        tbQuery.Text = String.Empty
        If Not (myCrmDataContext Is Nothing) Then
            Dim caption As String = TryCast(Me.cbQuery.SelectedItem, String)
            Dim query As CrmLinqQuery = Nothing
            Dim item As CrmLinqQuery
            For Each item In Me.queries
                If (item.Name = caption) Then
                    query = item
                    Exit For
                End If
            Next
            If (Not query Is Nothing) Then
                Me.dataGridView.DataSource = query.Query
                Me.tbQuery.Text = query.Description
            End If
        End If
    End Sub

    Friend Overrides Sub ControlsEnabled(ByVal state As Boolean)
        Me.cbQuery.Enabled = state
    End Sub

    Private Sub cbQuery_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cbQuery.SelectedIndexChanged
        If Me.queries IsNot Nothing Then
            Try
                RefreshQuery()
            Catch ex As Exception
                MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End If
    End Sub

    Private Sub dataGridView_DataError(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DataGridViewDataErrorEventArgs) Handles DataGridView.DataError

        If TryCast(e.Exception, ArgumentException) IsNot Nothing Then
            MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
        Else
            MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End If
    End Sub
End Class

Friend Class CrmLinqQuery
    Public Sub New(ByVal name As String, ByVal description As String, ByVal query As IQueryable)
        Me.Name = name
        Me.Description = description
        Me.Query = query
    End Sub

    Private qName As String
    Private qDescription As String
    Private qQuery As IQueryable

    Public Property Name() As String
        Get
            Return qName
        End Get
        Set(ByVal value As String)
            qName = value
        End Set
    End Property

    Public Property Description() As String
        Get
            Return Me.qDescription
        End Get
        Set(ByVal value As String)
            Me.qDescription = value
        End Set
    End Property

    Public Property Query() As IQueryable
        Get
            Return qQuery
        End Get
        Set(ByVal value As IQueryable)
            qQuery = value
        End Set
    End Property

    Public Overrides Function ToString() As String
        Return Me.Name
    End Function
End Class
